jitstats.lua

```diff
@@ -0,0 +1,462 @@
+    stats.traceflush = stats.traceflush + 1
+end
+
+function vmevent.texit(id, exitno)
+    debugprint("EXIT(%d): exit = %d", id, exitno)
+    stats.exits = stats.exits + 1
+end
+
+local function vmeventcb(what, ...)
+    local cb = vmevent[what]
+    if cb then 
+        cb(...)
+    end
+end
+
+function jitstats.start()
+    if jitstats.isrunning then
+        return false
+    end
+
+    jit.attach(vmeventcb, "trace")
+    jit.attach(vmevent.texit, "texit")
+    jitstats.isrunning = true
+    return true
+end
+
+function jitstats.stop()
+    if not jitstats.isrunning then
+        return false
+    end
+
+    jit.attach(vmeventcb)
+    jit.attach(vmevent.texit)
+    jitstats.isrunning = false
+    return true
+end
+
+local last_starts, last_exits
+
+function jitstats.reset(statstbl)
+    statstbl = statstbl or stats
+    statstbl.starts = 0
+    statstbl.aborts = 0
+    statstbl.abort_counts = {}
+    statstbl.completed = 0
+    statstbl.linktypes = {}
+    statstbl.stitch_starts = 0
+    statstbl.stitch_aborts = 0
+    statstbl.stitch_completed = 0
+    statstbl.side_starts = 0
+    statstbl.side_aborts = 0  
+    statstbl.side_completed = 0
+    statstbl.nyi_bc = {}
+    statstbl.exits = 0
+    statstbl.traceflush = 0
+    statstbl.func_blacklisted = 0
+    statstbl.loop_blacklisted = 0
+end
+
+local function table_copy(src, dst)
+    dst = dst or {}
+    for k,v in pairs(src) do
+        if type(v) ~= "table" then
+            dst[k] = v
+        else
+            dst[k] = table_copy(v, dst[k])
+        end
+    end
+    return dst
+end
+
+local function initorclear(tab, key)
+    tab[key] = tab[key] or {}
+    table.clear(tab[key])
+end
+
+--Optionally pass in a table to copy to snapshot to instead of creating a new table
+function jitstats.getsnapshot(dst)
+    if dst then
+        initorclear(dst, "nyi_bc")
+        initorclear(dst, "abort_counts")
+        initorclear(dst, "linktypes")
+    end
+
+    return table_copy(stats, dst)
+end
+
+local function diffstatbl(old, new, result, key)
+    assert(old[key], "missing old sub table")
+    assert(new[key], "missing new sub table")
+    assert(result[key], "missing old sub table")
+    old = old[key]
+    new = new[key]
+    result = result[key]
+
+    for k, v in pairs(new) do
+        local change = v - (old[k] or 0)
+        if change ~= 0 then
+            result[k] = change
+        end
+    end
+    return result
+end
+
+-- Get the diff between two jit stat snapshots
+function jitstats.diffsnapshots(snap1, snap2)
+    local result = {
+        abort_counts = {},
+        nyi_bc = {},
+        linktypes = {},
+    }
+
+    for k, v in pairs(snap2) do   
+        if type(v) == "number" then
+            result[k] = v - (snap1[k] or 0)
+        end
+    end
+
+    diffstatbl(snap1, snap2, result, "abort_counts")
+    diffstatbl(snap1, snap2, result, "nyi_bc")
+    diffstatbl(snap1, snap2, result, "linktypes")
+
+    return result
+end
+
+local last_starts, last_exits
```

Comment:
these look like private vars.  reduce scope?

Reply:
Now removed since function they were used in function didn't make sense was redesigned not to use them.